hudi-client/hudi-client-common/src/main/java/org/apache/hudi/common/HoodieEngineContext.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common;
+
+import org.apache.hudi.client.TaskContextSupplier;
+import org.apache.hudi.common.config.SerializableConfiguration;
+
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Base class contains the context information needed by the engine at runtime. It will be extended by different
+ * engine implementation if needed.
+ */
+public class HoodieEngineContext {
+  /**
+   * A wrapped hadoop configuration which can be serialized.
+   */
+  private SerializableConfiguration hadoopConf;
+
+  private TaskContextSupplier taskContextSupplier;
+
+  public HoodieEngineContext(SerializableConfiguration hadoopConf, TaskContextSupplier taskContextSupplier) {
+    this.hadoopConf = hadoopConf;
+    this.taskContextSupplier = taskContextSupplier;
+  }
+
+  public SerializableConfiguration getHadoopConf() {
+    return hadoopConf;
+  }
+
+  public TaskContextSupplier getTaskContextSupplier() {
+    return taskContextSupplier;
+  }
+
+  public <I, O> List<O> map(List<I> data, Function<I, O> func) {
```

Comment:
I think we should leave this abstract and let the engines implement this?  even for Java. Its better to have a `HoodieJavaEngineContext`. From what I can see, this is not overridden in `HoodieSparkEngineContext` and thus we lose the parallel execution that we currently have with Spark with this change.

Reply:
Also these APIs should take in a `parallelism` parameter, no?

Reply:
> I think we should leave this abstract and let the engines implement this? even for Java. Its better to have a `HoodieJavaEngineContext`. From what I can see, this is not overridden in `HoodieSparkEngineContext` and thus we lose the parallel execution that we currently have with Spark with this change.

as we discussed before, parallelDo model need a function as input parameter, Unfortunately, different engines need different type function, its hard to align them in an abstract parallelDo method. so we agree to use ` java.util.function.Function` as the unified input function. in this way, there is no need to distinguish spark and flink, and the parallelism is not needed too.

Reply:
@wangxianghu functionality wise, you are correct. it can be implemented just using Java. but, we do parallelization of different pieces of code e.g deletion of files in parallel using spark for a reason. It significantly speeds these up, for large tables. 

All I am saying is to implement the `HoodieSparkEngineContext#map` like below

```
 public <I, O> List<O> map(List<I> data, Function<I, O> func, int parallelism) {
    return javaSparkContext.parallelize(data, parallelism).map(func).collect();
 }
```

similarly for the other two methods. I don't see any issues with this. do you?

Reply:
> @wangxianghu functionality wise, you are correct. it can be implemented just using Java. but, we do parallelization of different pieces of code e.g deletion of files in parallel using spark for a reason. It significantly speeds these up, for large tables.
> 
> All I am saying is to implement the `HoodieSparkEngineContext#map` like below
> 
> ```
>  public <I, O> List<O> map(List<I> data, Function<I, O> func, int parallelism) {
>     return javaSparkContext.parallelize(data, parallelism).map(func).collect();
>  }
> ```
> 
> similarly for the other two methods. I don't see any issues with this. do you?

I know what you mean.

Reply:
Is it possible to take a `java.util.function.Function` and then within `HoodieSparkEngineContext#map` wrap that into a `org.apache.spark.api.java.function.Function` ?

Reply:
> Is it possible to take a `java.util.function.Function` and then within `HoodieSparkEngineContext#map` wrap that into a `org.apache.spark.api.java.function.Function` ?

let me try

Reply:
In Spark, there is a functional interface defined like this

```
package org.apache.spark.api.java.function;

import java.io.Serializable;

/**
 * Base interface for functions whose return types do not create special RDDs. PairFunction and
 * DoubleFunction are handled separately, to allow PairRDDs and DoubleRDDs to be constructed
 * when mapping RDDs of other types.
 */
@FunctionalInterface
public interface Function<T1, R> extends Serializable {
  R call(T1 v1) throws Exception;
}
```

Reply:
when the use passes in a regular lambda, into `rdd.map()`, this is what it gets converted into

Reply:
> when the use passes in a regular lambda, into `rdd.map()`, this is what it gets converted into

The serializable issue can be solved by introducing a seriableFuncition to replace `java.util.function.Function`

```
public interface SerializableFunction<I, O> extends Serializable {
  O call(I v1) throws Exception;
}
```
`HoodieEngineContext` can be
```
public abstract class HoodieEngineContext {
  public abstract  <I, O> List<O> map(List<I> data, SerializableFunction<I, O> func, int parallelism) ;
}
```

`HoodieSparkEngineContext` can be
```
public class HoodieSparkEngineContext extends HoodieEngineContext {
  private static JavaSparkContext jsc;

  // tmp
  static {
    SparkConf conf = new SparkConf()
        .setMaster("local[4]")
        .set("spark.driver.host","localhost")
        .setAppName("HoodieSparkEngineContext");

    jsc = new JavaSparkContext(conf);
  }
  
  @Override
  public <I, O> List<O> map(List<I> data, SerializableFunction<I, O> func, int parallelism) {
    return jsc.parallelize(data, parallelism).map(func::call).collect();
  }
}
```
this works :)

Reply:
@wangxianghu This is awesome. Hopefully this can reduce the amount of code you need to write for Flink significantly. `TestMarkerFiles` seems to pass, so guess the serialization etc is working as expected. 

We can go ahead with doing more files in this approach and remerge the base/child classes back as much as possible. cc @leesf @yanghua  as well in case they have more things to add. 

cc @bvaradar as well as FYI

Reply:
> @wangxianghu This is awesome. Hopefully this can reduce the amount of code you need to write for Flink significantly. `TestMarkerFiles` seems to pass, so guess the serialization etc is working as expected.
> 
> We can go ahead with doing more files in this approach and remerge the base/child classes back as much as possible. cc @leesf @yanghua as well in case they have more things to add.
> 
> cc @bvaradar as well as FYI

Yes, it also reduce tons of code in the refactoring. I'm working on it ,hope to finish it today or tomorrow